src/mlpack/methods/ann/loss_functions/cross_entropy_error_impl.hpp

```diff
@@ -32,8 +32,14 @@ CrossEntropyError<InputDataType, OutputDataType>::Forward(
     const InputType& input,
     const TargetType& target)
 {
-  return -arma::accu(target % arma::log(input + eps) +
+  InputType loss =  -(target % arma::log(input + eps) +
       (1. - target) % arma::log(1. - input + eps));
+  typename InputType::elem_type lossSum = arma::accu(loss);
+
+  if (reduction)
+    return lossSum;
+
+  return lossSum / input.n_elem;
```

Comment:
I think if `reduction = true`, mean loss has to be returned.

Reply:
I am returning sum, when reduction = true.
That I am doing for every loss function, so there shouldn't be any non-uniformity, I guess.

Is there any specific reason why I should return mean instead of sum, when reduction = true?

Reply:
Not a big deal. But it just looked counter-intuitive. Any new user would expect that "reduced" form is less than "non-reduced" form. (It's like swapping the keys `Home` and `End` in your keyboard)

Reply:
The non-reduced form is a vector / matrix / cube.
For all the loss functions I have touched on above, the non-reduced loss is stored in the variable loss.
And the reduced form is a single number, either the sum or the mean of all the values in the loss variable.

The reason why I preferred the sum reduction is simply that sum will have a greater value than mean for all cases, which might be more effective when back propagating it through previous layers. 

Lets wait for some more opinions on this.
If more people feel this is counter-intuitive, I will swap them back to what you suggested :)

Reply:
If we do this for all the other functions, which I think you do, it's fine for me.